File: 977-distinct-subsequences-ii/distinct-subsequences-ii.js
const MOD = 1e9 + 7;

const distinctSubseqII = s => {
    let dp = new Int32Array(26), tot = 0;

    for (let i = 0; i < s.length; i++) {
        const c = s.charCodeAt(i) - 97;
        const add = (tot - dp[c] + MOD) % MOD;

        dp[c] = 1 + tot;
        tot = (dp[c] + add) % MOD;
    }

    return tot;
};

File: 977-distinct-subsequences-ii/distinct-subsequences-ii.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        MOD = 1000000007
        
        dp = 1
        last = [0] * 26
        
        for ch in s:
            index = ord(ch) - ord('a')
            
            old_dp = dp
            
            dp = (2 * dp - last[index] + MOD) % MOD
            
            last[index] = old_dp
        
        return (dp - 1 + MOD) % MOD